src/directory_and_file_reader.c

```diff
@@ -0,0 +1,289 @@
+#else
+  char * path_buf[RCUTILS_DIR_PATH_MAX];
+  TCHAR * pathp = &dir->path[strlen(dir->path) - 1];
+  while (pathp != dir->path && (*pathp == _TEXT('\\') || *pathp == _TEXT('/')))
+  {
+    *pathp = _TEXT('\0');
+    pathp++;
+  }
+
+  if (rcutils_snprintf(path_buf, "%s%s", dir->path, _TEXT("\\*")) < 0)
+  {
+    RCUTILS_SET_ERROR_MSG("failed concat strings");
+    return RCUTILS_RET_ERROR;
+  }
+
+  dir->dir = FindFirstFile(path_buf, &dir->f);
+  if (dir->dir == INVALID_HANDLE_VALUE) {
+#endif
+    int errsv = errno;
+    if (errsv == EACCES) {
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("permission denied '%s'", dir->path);
+    } else if (errsv == EBADF) {
+      RCUTILS_SET_ERROR_MSG("fd is not a valid file descriptor opened for reading.");
+    } else if (errsv == EMFILE) {
+      RCUTILS_SET_ERROR_MSG("Too many file descriptors in use by proces.");
+    } else if (errsv == ENFILE) {
+      RCUTILS_SET_ERROR_MSG("Too many files are currently open in the system.");
+    } else if (errsv == ENOENT) {
+      RCUTILS_SET_ERROR_MSG("Directory does not exist, or name is an empty string.");
+    } else if (errsv == ENOMEM) {
+      RCUTILS_SET_ERROR_MSG("Insufficient memory to complete the operation.");
+    } else if (errsv == ENOTDIR) {
+      RCUTILS_SET_ERROR_MSG("name is not a directory.");
+    }
+    dir->allocator.deallocate(dir->path, dir->allocator.state);
+    return RCUTILS_RET_ERROR;
+  }
+  dir->has_next = 1;
+
+#ifndef _WIN32
+  dir->f = readdir(dir->dir);
+  if (dir->f == NULL) {
+    dir->has_next = 0;
+  }
+#else
+  WIN32_FIND_DATA _temp_f;
+  if (!FindNextFile(dir->dir, &_temp_f)) {
+    dir->has_next = 0;
+  }
+#endif
+  return RCUTILS_RET_OK;
+}
+
+rcutils_ret_t
+rcutils_readfile(rcutils_dir_t * dir, rcutils_file_t * file)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(dir, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(dir->path, RCUTILS_RET_INVALID_ARGUMENT);
+#ifndef _WIN32
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(dir->f, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(dir->f->d_name, RCUTILS_RET_INVALID_ARGUMENT);
+#endif
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(file, RCUTILS_RET_INVALID_ARGUMENT);
+
+  const char * filename =
+#ifndef _WIN32
+    dir->f->d_name;
+#else
+    dir->f.cFileName;
+#endif
+
+  if (file->path != NULL) {
+    dir->allocator.deallocate(file->path, dir->allocator.state);
+  }
+  file->path = rcutils_strdup(dir->path, dir->allocator);
+  if (NULL == file->path) {
+    RCUTILS_SET_ERROR_MSG("unable to allocate memory");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+
+  if (file->name != NULL) {
+    dir->allocator.deallocate(file->name, dir->allocator.state);
+  }
+  file->name = rcutils_strdup(filename, dir->allocator);
+  if (NULL == file->name) {
+    dir->allocator.deallocate(file->name, dir->allocator.state);
+    RCUTILS_SET_ERROR_MSG("unable to allocate memory");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+
+#ifndef _WIN32
+  if (lstat(
+#else
+  if (_tstat(
+#endif
```

Comment:
Actually, this stanza is likely the issue... I'd recommend using a `#define` for this function name if that's all that really is different.